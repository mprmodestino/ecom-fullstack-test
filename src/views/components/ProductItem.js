import React from "react";
import PropTypes from "prop-types";

export function formatNumber(num) {
  return (Number(num) / 100).toFixed(2);
}

export default function ProductItem({ item }) {
  const {
    title,
    description,
    image,
    price,
    currency,
    priceLabel,
    productLabel,
    cta,
    ctaLink
  } = item;

  return (
    <article className="c-product-item">
      <header className="c-product-item__header">
        {productLabel ? (
          <span
            className={`c-product-item__label c-product-item__label--${productLabel}`}
          >
            {productLabel}
          </span>
        ) : null}
        <img
          className="c-product-item__image"
          src={image.path}
          alt={image.alt}
        />
      </header>
      <div className="c-product-item__copy">
        <h3 className="c-product-item__title">{title}</h3>
        <p className="c-product-item__description">{description}</p>
        <p className="c-product-item__pricecontainer">
          <span className="c-product-item__pricelabel">{priceLabel}</span>{" "}
          <span className="c-product-item__price">
            {currency}
            {formatNumber(price)}
          </span>
        </p>
        <a className="c-product-item__link" href={ctaLink}>
          <strong>{cta}</strong>
        </a>
      </div>
    </article>
  );
}

ProductItem.propTypes = {
  product: PropTypes.shape({
    title: PropTypes.string,
    description: PropTypes.string,
    image: PropTypes.shape({
      path: PropTypes.string,
      alt: PropTypes.string
    }),
    price: PropTypes.number,
    currency: PropTypes.string,
    priceLabel: PropTypes.string,
    productLabel: PropTypes.string,
    cta: PropTypes.string,
    ctaLink: PropTypes.string
  })
};
