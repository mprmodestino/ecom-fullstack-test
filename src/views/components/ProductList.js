import React from "react";
import ProductItem from "./ProductItem";

export default function ProductList({ items }) {
  return (
    <div className="u-container">
      <section className="c-product-list">
        {items.map((item, index) => <ProductItem key={index} item={item} />)}
      </section>
    </div>
  );
}
