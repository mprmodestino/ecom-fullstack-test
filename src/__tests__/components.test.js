import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-15";
import toJson from "enzyme-to-json";
import ProductList from "../views/components/ProductList";
import ProductItem from "../views/components/ProductItem";

Enzyme.configure({ adapter: new Adapter() });

describe("components", () => {
  describe("Product List", () => {
    it("should render self", () => {
      const enzymeWrapper = shallow(<ProductList items={[]} />);
      expect(enzymeWrapper.find("section").hasClass("c-product-list")).toBe(
        true
      );
      expect(enzymeWrapper.find("ProductItem").length).toEqual(0);
    });
    it("should render self and subcomponents", () => {
      const items = [
        {
          title: "Simple Canvas",
          description: "Lets your pictures speak for themselves.",
          image: {
            path: "/images/product.jpg",
            alt: "Simple Canvas"
          },
          price: 1500,
          currency: "£",
          priceLabel: "From",
          productLabel: "bestseller",
          cta: "Shop Now",
          ctaLink: "/random/link/to/no/where"
        },
        {
          title: "Collage Canvas",
          description:
            "Can't choose just one pic? Put your favourite photos on one canvas - a collage.",
          image: {
            path: "/images/product.jpg",
            alt: "Collage Canvas"
          },
          price: 2500,
          currency: "£",
          priceLabel: "From",
          productLabel: "",
          cta: "Shop Now",
          ctaLink: "/random/link/to/no/where"
        }
      ];
      const enzymeWrapper = shallow(<ProductList items={items} />);
      expect(enzymeWrapper.find("ProductItem").length).toEqual(2);
      expect(toJson(enzymeWrapper)).toMatchSnapshot();
    });
  });
  describe("Product Item regular", () => {
    const item = {
      title: "Simple Canvas",
      description: "Lets your pictures speak for themselves.",
      image: {
        path: "/images/product.jpg",
        alt: "Simple Canvas"
      },
      price: 1500,
      currency: "£",
      priceLabel: "From",
      cta: "Shop Now",
      ctaLink: "/random/link/to/no/where"
    };
    const enzymeWrapper = shallow(<ProductItem item={item} />);

    it("should render self", () => {
      expect(enzymeWrapper.find("article").hasClass("c-product-item")).toBe(
        true
      );
    });
    it("should render title", () => {
      expect(enzymeWrapper.find(".c-product-item__title").length).toEqual(1);
      expect(enzymeWrapper.find(".c-product-item__title").text()).toBe(
        "Simple Canvas"
      );
    });
    it("should render description", () => {
      expect(enzymeWrapper.find(".c-product-item__description").length).toEqual(
        1
      );
      expect(enzymeWrapper.find(".c-product-item__description").text()).toBe(
        "Lets your pictures speak for themselves."
      );
    });
    it("should render price", () => {
      expect(
        enzymeWrapper.find(".c-product-item__pricecontainer").length
      ).toEqual(1);
      expect(enzymeWrapper.find(".c-product-item__pricecontainer").text()).toBe(
        `From £${(Number(1500) / 100).toFixed(2)}`
      );
    });
    it("should render link", () => {
      expect(enzymeWrapper.find(".c-product-item__link").length).toEqual(1);

      expect(enzymeWrapper.find(".c-product-item__link").text()).toBe(
        `Shop Now`
      );
    });
    it("should render image", () => {
      expect(enzymeWrapper.find(".c-product-item__image").length).toEqual(1);
    });
  });
  describe("Product Item best seller", () => {
    const item = {
      title: "Simple Canvas",
      description: "Lets your pictures speak for themselves.",
      image: {
        path: "/images/product.jpg",
        alt: "Simple Canvas"
      },
      price: 1500,
      currency: "£",
      priceLabel: "From",
      productLabel: "bestseller",
      cta: "Shop Now",
      ctaLink: "/random/link/to/no/where"
    };
    const enzymeWrapper = shallow(<ProductItem item={item} />);
    it("should render item label", () => {
      expect(enzymeWrapper.find(".c-product-item__label").text()).toBe(
        "bestseller"
      );
    });
  });
});
