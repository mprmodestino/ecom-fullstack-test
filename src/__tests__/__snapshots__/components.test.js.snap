// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Product List should render self and subcomponents 1`] = `
<div
  className="u-container"
>
  <section
    className="c-product-list"
  >
    <ProductItem
      item={
        Object {
          "cta": "Shop Now",
          "ctaLink": "/random/link/to/no/where",
          "currency": "£",
          "description": "Lets your pictures speak for themselves.",
          "image": Object {
            "alt": "Simple Canvas",
            "path": "/images/product.jpg",
          },
          "price": 1500,
          "priceLabel": "From",
          "productLabel": "bestseller",
          "title": "Simple Canvas",
        }
      }
      key="0"
    />
    <ProductItem
      item={
        Object {
          "cta": "Shop Now",
          "ctaLink": "/random/link/to/no/where",
          "currency": "£",
          "description": "Can't choose just one pic? Put your favourite photos on one canvas - a collage.",
          "image": Object {
            "alt": "Collage Canvas",
            "path": "/images/product.jpg",
          },
          "price": 2500,
          "priceLabel": "From",
          "productLabel": "",
          "title": "Collage Canvas",
        }
      }
      key="1"
    />
  </section>
</div>
`;
